### DISCUSSION TITLE
[10일차-지원] promise와 callback 차이를 알려주세요.

### DISCUSSION THREAD

**Discussion by smarfy99** [zzangCS/zzangInterview#36] [프론트 면접 질문]
promise와 callback의 차이를 알려주세요!

**Comment by ez1n** [comment]
- **비동기 처리**
특정 코드의 실행이 완료될 때까지 기다리지 않고 다음 코드를 먼저 수행하는 자바스크립트의 특성
    
- **promise**
자바스크립트 비동기 처리에 사용되는 객체
대기(pending), 이행(fulfilled), 실패(rejected) 세 가지의 상태를 가지며 처리 완료된 결과값을 객체에 저장하여 받을 수 있다. (코드 작성이 용이해짐)
    
- **callback**
비동기 로직의 결과값을 처리하기 위해 callback 안에서만 처리 해야하며, callback 밖에서는 비동기에서 온 결과값을 알 수 없다.

**Comment by skyhanull** [comment]
callback
* 자바스크립트에서 비동기를 처리하기 위해 나옴
* 함수 내부에서만 사용이 되며 연속해서 사용 시 callback hell이 나올 수 있다

promise
* callback hell을 해결하기 위해 나온 문법
* pullfilled,resolved,rejected 3가지 상태를 가지며 함수 외부에서도 사용이 가능하다?
* 하지만 이또한 체이닝으로 hell이 발생할 수 있다

**Comment by smarfy99** [comment]
비동기 처리
- 특정 코드의 실행이 완료될 때까지 기다리지 않고 다음 코드를 먼저 수헹하는 것

콜백
- 비동기를 콜백 안에서만 처리하고, 콜백 밖에서는 비동기에서 온 값을 알 수 없다
- 콜백 지옥이 생긴다

프로미스
- return value를 이용할 수 있다
- error handling이 동기식 코드와 유사하게 쓰일 수 있다
- 상태에 따라 다르게 처리해줄 수 있다.

**Reply by smarfy99** [reply]
https://velog.io/@wngud4950/%ED%94%84%EB%A1%9C%EB%AF%B8%EC%8A%A4Promise%EB%9E%80-%EC%BD%9C%EB%B0%B1%ED%95%A8%EC%88%98%EC%99%80%EC%9D%98-%EC%B0%A8%EC%9D%B4%EC%A0%90
